### PR TITLE
introduce 'fatal assertions'

### PR DESCRIPTION
**Pull request by kroening**
This introduces a variant of `ASSERT` instructions that are fatal when they are refuted.  Execution paths through fatal assertions that are refuted are undefined.  Assertions that (otherwise) pass and that are reachable from a refuted fatal assertion are now reported as `UNKNOWN`.

The motivating use-case for fatal assertions is undefined behavior in languages such as C/C++ or Rust.

- [X] Each commit message has a non-empty body, explaining why the change was made.
- [X] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [X] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [X] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Comment by peterschrammel** [comment]
Is [UNKNOWN](https://github.com/diffblue/cbmc/blob/develop/src/goto-checker/properties.h#L29) the right status here? Or should there are rather be a new status UNDEFINED?

**Comment by kroening** [comment]
> Is [UNKNOWN](https://github.com/diffblue/cbmc/blob/develop/src/goto-checker/properties.h#L29) the right status here? Or should there are rather be a new status UNDEFINED?

I did think about revising the status outputs to increase clarity. But that should happen independently, as a separate discussion.

An issue with "UNDEFINED" is that the word "undefined behavior" is very specific to C/C++, and we may wish our status codes to be more general. Also, it may mislead a reader to believe that there is undefined behavior at the given location. There may well be, but the point is that the undefined behavior is _before_ that assertion in program order.

**Comment by zhassan-aws** [comment]
In Kani, we use UNDETERMINED.

**Comment by kroening** [comment]
I've pulled out the question of division-by-zero for floating-point numbers into a separate PR: #8233

**Comment by celinval** [comment]
Are fatal assertions encoded as assert + assume? How customizable will this new assertion be? Will users also be able to instrument their code with fatal assertions?

**Comment by kroening** [comment]
> Are fatal assertions encoded as assert + assume?

I am afraid that would yield the opposite result: you'd be more likely to get SUCCESS for the downstream properties, making the user believe that they are proven when really they are not.

> How customizable will this new assertion be? Will users also be able to instrument their code with fatal assertions?

We could certainly consider a `__CPROVER_fatal_assert`, or the like.

**Comment by celinval** [comment]
> > Are fatal assertions encoded as assert + assume?
> 
> I am afraid that would yield the opposite result: you'd be more likely to get SUCCESS for the downstream properties, making the user believe that they are proven when really they are not.
> 

In Kani, we over-approximate by marking all passing properties as undetermined.

**Comment by tautschnig** [comment]
> > > Are fatal assertions encoded as assert + assume?
> > 
> > 
> > I am afraid that would yield the opposite result: you'd be more likely to get SUCCESS for the downstream properties, making the user believe that they are proven when really they are not.
> 
> In Kani, we over-approximate by marking all passing properties as undetermined.

#8242 now implements the assert + assume when users request this.

**Comment by kroening** [comment]
This now marks refuted assertions after refuted fatal assertions as UNKNOWN.

**Comment by kroening** [comment]
> This now marks refuted assertions after refuted fatal assertions as UNKNOWN.

This comes at a price: to identify those assertions, the trace needs to be computed. That trace computation is broken in a number of cases, now marked as KNOWNBUG. Note that these aren't regressions caused by this PR, but just exposed by this PR.

We can try to fix these beforehand, and delay this PR. I believe that the step towards reporting sound results is very important, and that we should do it for 6.0.

**Comment by tautschnig** [comment]
> This comes at a price: to identify those assertions, the trace needs to be computed. That trace computation is broken in a number of cases, now marked as KNOWNBUG. Note that these aren't regressions caused by this PR, but just exposed by this PR.

What kind of brokenness is this? I have previously only seen that computing traces can be very expensive (see #5714).

**Comment by kroening** [comment]
> What kind of brokenness is this? I have previously only seen that computing traces can be very expensive (see #5714).

It seems to be a miscellaneous set of issues that get triggered by calling `get(...)`; some of them in the contract code, and also one in the interface to cprover_smt.

I didn't see a case of things getting very slow, but I'd say that's just as bad as throwing an error. This ought to be linear.

File: regression/symtab2gb/multiple_symtabs/test.desc [hunk 1]
```diff
@@ -7,4 +7,3 @@ user.json_symtab library.json_symtab
 ^\[2\] file library.adb line \d+ assertion: FAILURE$
 ^VERIFICATION FAILED$
 --
-error
```
Comment (tautschnig) [inline regression/symtab2gb/multiple_symtabs/test.desc]: Why did this line get removed?

File: regression/symtab2gb/single_symtab/test.desc [hunk 2]
```diff
@@ -7,4 +7,3 @@ entry_point.json_symtab
 ^\[2\] file entry_point.adb line \d+ assertion: SUCCESS$
 ^VERIFICATION FAILED$
 --
-error
```
Comment (tautschnig) [inline regression/symtab2gb/single_symtab/test.desc]: Same here: Why did this line get removed?

File: src/goto-checker/fatal_assertions.cpp [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+  while(!working.empty())
+  {
+    auto loc = working.top();
+    working.pop();
+
+    auto insertion_result = fixpoint.insert(loc);
+    if(!insertion_result.second)
+      continue; // seen already
+
+    if(loc.target->is_function_call())
```
Comment (remi-delmas-3000) [inline src/goto-checker/fatal_assertions.cpp:81]: does this assume that function pointer calls have been eliminated ?

File: src/goto-checker/fatal_assertions.cpp [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+        working.emplace(loc.function_it, successor);
+    }
+  }
+
+  return fixpoint;
+}
+
+/// Proven assertions after refuted fatal assertions
+/// are marked as UNKNOWN.
+void propagate_fatal_to_proven(
```
Comment (remi-delmas-3000) [inline src/goto-checker/fatal_assertions.cpp:114]: why keep refuted assertions after refuted fatal assertions intact ? The counter example could also be spurious or not worth investigating.

File: src/goto-checker/fatal_assertions.cpp [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+  propertiest &properties,
+  const goto_functionst &goto_functions)
+{
+  // Iterate to find refuted fatal assertions. Anything reachalble
+  // from there is a 'fatal loc'.
+  loc_sett fatal_locs;
```
Comment (remi-delmas-3000) [inline src/goto-checker/fatal_assertions.cpp:120]: Is there a risk to flip the status of a fatal assertion from FAIL to UNKOWN if the goto model contains mutually recursive functions  ? (a fatal assertion could appear to be after itself ?)

**Review by remi-delmas-3000 — COMMENTED** [review]
Maybe add tests with mutually recursive functions containing fatal assertions

**Review by remi-delmas-3000 — COMMENTED** [review]
Is there a documentation update for fatal assertions  ?

Are there ways to invoke CBMC which would result in the propagation not happening ?
i.e. is `all_properties_verifier_with_trace_storaget` the only verifier that users can invoke through the CLI ?

File: regression/symtab2gb/multiple_symtabs/test.desc (hunk 1 above)
Comment (kroening) [reply regression/symtab2gb/multiple_symtabs/test.desc]: The test triggered on "error trace", i.e., fails when `--trace` is added.

File: src/goto-checker/fatal_assertions.cpp (hunk 3 above)
Comment (kroening) [reply src/goto-checker/fatal_assertions.cpp:81]: Yes.

File: src/goto-checker/fatal_assertions.cpp (hunk 4 above)
Comment (kroening) [reply src/goto-checker/fatal_assertions.cpp:114]: Note the discussion on the PR. The problem is that enabling the computation of counterexamples is expensive and doesn't work. Hence I've taken that out again.

File: src/goto-checker/fatal_assertions.cpp (hunk 5 above)
Comment (kroening) [reply src/goto-checker/fatal_assertions.cpp:120]: No, assertions that are failed simply stay failed.

**Comment by kroening** [comment]
> Is there a documentation update for fatal assertions ?

This is a bigger job. It will need to be done eventually, but I'd do it as part of a bigger doc overhaul. It requires explaining what UB is, why it's bad, and why it's not locally contained.

> Are there ways to invoke CBMC which would result in the propagation not happening ?

Not that I know of.

File: regression/symtab2gb/multiple_symtabs/test.desc (hunk 1 above)
Comment (tautschnig) [reply regression/symtab2gb/multiple_symtabs/test.desc]: But ... this PR no longer adds a need to compute traces?

Comment (kroening) [reply regression/symtab2gb/multiple_symtabs/test.desc]: Yes, but our future selves still want this. But could pull over into separate PR.

Comment (tautschnig) [reply regression/symtab2gb/multiple_symtabs/test.desc]: I'm sorry, I had to convince myself of the problem... Done and now PR'ed as https://github.com/diffblue/cbmc/pull/8270.

**Review by tautschnig — APPROVED** [review]
Approving, but would appreciate if https://github.com/diffblue/cbmc/pull/8270 could be merged first to have a proper separation of concerns.

**Pull request merged**